### PR TITLE
Synapse trace

### PR DESCRIPTION
**Pull request by saeedark**
This PR have many changes, the main one will move trace behaviour into synapses instead of neurons with allows for more flexibility.

### CODE REVIEW

File: conex/helpers/transforms/masks.py [hunk 1]
```diff
@@ -33,7 +35,20 @@ def __call__(self, img):
         )
         for index, ij in enumerate(product(range(self.m), range(self.n))):
             i, j = ij
-            result.append(TF.erase(img, i * h_grid, j * w_grid, h_grid, w_grid, v=0))
+            h_cor = i * h_grid + self.gap[0]
```
Comment (realamirhe) [inline conex/helpers/transforms/masks.py:38]: deconstruct the gap, and use local variables

```py
gap_left, gap_bottom, gap_up, gap_right = self.gap
```

or define constants

```py
class Direction:
   LEFT = 0
   BOTTOM = 1
   UP = 2
   RIGHT = 3

h_cor = i * h_grid + self.gap[Direction.LEFT]

```

File: conex/behaviors/synapses/learning.py [hunk 2]
```diff
@@ -249,14 +215,14 @@ def initialize(self, synapse):
 
         # messy till I move trace to synapse.
         pre_tau = [
-            synapse.src.behavior[key_behavior]
-            for key_behavior in synapse.src.behavior
-            if isinstance(synapse.src.behavior[key_behavior], SpikeTrace)
+            synapse.behavior[key_behavior]
+            for key_behavior in synapse.behavior
```
[the remark this answers is not in the document]
Comment (saeedark) [reply conex/behaviors/synapses/learning.py:219]: This specific learning method needs variable tau_s to perform its operation; Since the trace now resides on synapse accessing the src neuron group is wrong.

File: conex/behaviors/synapses/specs.py [hunk 3]
```diff
@@ -288,3 +291,93 @@ def forward(self, synapses):
...
+    Calculates the pre synaptic spike trace.
+
+    Note : should be placed after spike catcher behavior.
+
+    Args:
+        tau_s (float): decay term for spike trace. The default is None.
+        spike_scale (float): the increase effect of spikes on the trace.
+    """
+
+    def __init__(self, tau_s, *args, spike_scale=1.0, **kwargs):
```
[the remark this answers is not in the document]
Comment (saeedark) [reply conex/behaviors/synapses/specs.py:337]: As far as I know, using the symbol "/" function signature prevents specifying "tau_s" by keyword which is not desired.

File: conex/helpers/transforms/masks.py [hunk 4]
```diff
@@ -15,12 +15,14 @@ class GridEraseMask:
         m (int): The grid row size.
         n (int): The grid column size.
         random (bool): If true, shuffles the order of the masks.
+        gap (tuple(int)): left, bottom, up, right gaps for the cell.
```
[the remark this answers is not in the document]
Comment (saeedark) [reply conex/helpers/transforms/masks.py]: done, too bad vim keybinding are not considered universal

File: conex/helpers/transforms/masks.py (hunk 1 above)
Comment (saeedark) [reply conex/helpers/transforms/masks.py]: defining a class is overkill, deconstructing is pretty reasonable that a new commit should've elected this style.

File: conex/behaviors/synapses/specs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (realamirhe) [reply conex/behaviors/synapses/specs.py:337]: what about forcing a named keyword argument for the `tau_s`

```python
# current arguments
PreTrace(0.1, 1,2,3,4,5,6, spike_scale=1, a=1,b=2,c=3) 
# with changes
def __init__(self, tau_s, /, *, spike_scale=1.0, **kwargs):
PreTrace(0.1, 1,2,3,4,5,6, spike_scale=1, a=1,b=2,c=3) 
              ^^^^^^^^^^^ issue 🐛
```

I don't want to be stuck within this issue, make it named, or resolve it for project's current stage.

File: conex/helpers/transforms/masks.py [hunk 5]
```diff
@@ -15,25 +15,41 @@ class GridEraseMask:
...
 
         result = []
         location = torch.ones(
             self.m * self.n, self.m, self.n, dtype=torch.bool, device=img.device
         )
         for index, ij in enumerate(product(range(self.m), range(self.n))):
             i, j = ij
-            result.append(TF.erase(img, i * h_grid, j * w_grid, h_grid, w_grid, v=0))
+            h_cor = i * h_grid + gap_left
+            dh = h_cor if h_cor < 0 else 0
```
Comment (realamirhe) [inline conex/helpers/transforms/masks.py:40]:
```suggestion
            dh = min(h_cor, 0) 
```

File: conex/helpers/transforms/masks.py [hunk 6]
```diff
@@ -15,25 +15,41 @@ class GridEraseMask:
...
+            w_cor = j * w_grid + gap_top
+            dw = w_cor if w_cor < 0 else 0
```
Comment (realamirhe) [inline conex/helpers/transforms/masks.py:42]:
```suggestion
            dw = min(w_cor, 0)
```

**Review by realamirhe — APPROVED** [review]
